### ISSUE TITLE
llvm-mca, good example to start profiling

### ISSUE THREAD

**Issue by xshen053** [llvm/llvm-project#63716]
Hi, If I want to do profiling to analyze some bottleneck using llvm-mca on arm64 (mac m1), is there any good example assembly code recommended to start with? Cuz I think if I put a bunch number of code, probably I am not able to analyze it well!

**Comment by xshen053** [comment]
In this video, https://www.youtube.com/watch?v=Ku2D8bjEGXk
The speaker give an analysis on x86 based on a doc product which only has 3 lines of code.
I want to do such analyisis on arm64, so I want to get some advices, what kind of code can I pick.

**Comment by RKSimon** [comment]
I'm not sure what is the closest available CPU to analyze for m1, but this should get you going:

https://rust.godbolt.org/z/naYzrYEjv

**Comment by xshen053** [comment]
> 

Thank you so much for this link!!

Currently M1 (they're using Apple A7 "Cyclone"'s scheduling model):
I am not sure if there is any difference between A7 and M1.
But I will start analyzing m1 using A7 scheduling model, could I ask you another question?

if I use the closest available CPU, Is the data generate by simulation correct? If not, how can I know what's the deviation. Cuz if I understand correctly, even if it's the closest cpu, as long as it's not the same, it will bring some deviation?

**Comment by RKSimon** [comment]
OK - try setting the llvm-mca arguments to ```-mtriple=aarch64 -mcpu=cyclone```

Given straight line code the simulation will give you an approximate estimation for throughput / dependencies etc. but it depends on the accuracy of the cyclone scheduler model, and often many subtleties can be missed. Plus the sim has to make a lot of assumptions (notably memory latency etc.). llvm-mca also makes no attempt at instruction decode simulation, although that shouldn't be a big problem for arm cpus.

**Comment by mshockwave** [comment]
> if I use the closest available CPU, Is the data generate by simulation correct? If not, how can I know what's the deviation. Cuz if I understand correctly, even if it's the closest cpu, as long as it's not the same, it will bring some deviation?

I'm not sure if it runs on MacOSX (as it relies on Perf AFAIK), but [`llvm-exegesis`](https://releases.llvm.org/7.0.0/docs/CommandGuide/llvm-exegesis.html) can be used to find the deviation between LLVM's scheduling information and the processor characteristics on real hardware.

**Comment by xshen053** [comment]
Thank you so much.

I finally decided to start analyzing graviton neoverse n2 instead, because llvm-mca has neoverse n2! If we are using exact the same model, **is it still possible to have some deviation?**

**Comment by mshockwave** [comment]
> **is it still possible to have some deviation?**

Absolutely, as Simon pointed out in earlier comments, there are features not covered by MCA such as processor frontend (i.e. instruction fetch, I$, branch predictor), precise simulations on memory operations (e.g. memory reordering, D$), and SMT (aka HyperThreading).
If you're interested in this topic, feel free to checkout papers about throughput analysis -- that's the name of the field.

**Comment by xshen053** [comment]
> 



> > **is it still possible to have some deviation?**
> 
> Absolutely, as Simon pointed out in earlier comments, there are features not covered by MCA such as processor frontend (i.e. instruction fetch, I$, branch predictor), precise simulations on memory operations (e.g. precise memory reordering, D$), and SMT (aka HyperThreading). If you're interested in this topic, feel free to checkout papers about throughput analysis -- that's the name of the field.

Thank you so much!!
But if I want to observe some difference, eg. resource pressure between following two situation
<img width="911" alt="image" src="https://github.com/llvm/llvm-project/assets/97472036/45ba930c-0d88-48b7-b720-4ba9df41bbda">
like before loop unrolling and after loop unrolling, I assume the current model is enough to achieve this goal?

**Comment by RKSimon** [comment]
yes, you should be able to use the "-iterations=XXX" argument to control the iteration count to investigate loop unrolling
